### PR TITLE
Assign project draft publish statuses

### PR DESCRIPTION
**Pull request by chamaliperera512**

### CODE REVIEW

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 1]
```diff
@@ -302,14 +302,16 @@ class ProjectController {
...
-             ecoScienceTypes: ecoScienceTypes
+            def projectActivities = projectActivityService.getAllByProject(project.projectId, "docs", params?.version, true)
+
+            [project        : project,
+             siteDocuments  : siteInfo.documents ?: '[]',
+             site           : siteInfo.site,
+             programs       : metadataService.programsModel(),
+             scienceTypes   : scienceTypes,
+             ecoScienceTypes: ecoScienceTypes,
+             activities     : projectActivities
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:313]: `activities` needs renaming.

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 2]
```diff
@@ -826,6 +854,11 @@ class ProjectController {
         }
 
         if (trimmedParams.isUserPage) {
+            boolean isAlaAdmin = userService.userIsAlaAdmin()
+
+            if (isAlaAdmin)
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:859]: It looks to me ALA admin can only see published projects in their user page. is the logic correct?

File: grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy [hunk 3]
```diff
@@ -909,7 +909,8 @@ class ProjectService {
                         index = facets.size()
                         facets.add(index, specialFacet.clone())
                     } else {
-                        facets.putAt(index, specialFacet.clone())
+                        int facetIndex = facets.size()
```
Comment (temi) [inline grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy:912]: Any reason why this code has changed? It is supposed to replace special facets at the index found in `hubFacets`.

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy (hunk 2 above)
Comment (temi) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:859]: The code needs to satisfy the following conditions
1. ALA admin can see published and unpublished projects on project finder.
2. Regular users should only see published projects only.
3. Regular users when viewing `My projects` should be able to see their projects that are both published and unpublished.

**Review by temi — CHANGES_REQUESTED** [review]
Reviewed. I am available to discuss further.

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 4]
```diff
@@ -825,7 +872,17 @@ class ProjectController {
             trimmedParams.status = null
         }
 
+        boolean isAlaAdmin = userService.userIsAlaAdmin()
+
+        if (!isAlaAdmin)
+            fq.push('projLifecycleStatus:published')
+
         if (trimmedParams.isUserPage) {
+            if (!fq.contains('projLifecycleStatus:published'))
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:881]: Are you trying to get both published and unpublished here? Not adding any `fq` will get both projects.

**Review by temi — CHANGES_REQUESTED** [review]
I am happy to discuss this further.

File: grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy [hunk 5]
```diff
@@ -908,9 +908,13 @@ class ProjectService {
                     if(index >= facets?.size()){
                         index = facets.size()
                         facets.add(index, specialFacet.clone())
-                    } else {
-                        int facetIndex = facets.size()
-                        facets.putAt(facetIndex, specialFacet.clone())
+                    }
+                    else {
+                        def item =  hubFacets[index - 1]
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy]: Handled

File: grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy [hunk 6]
```diff
@@ -908,9 +908,13 @@ class ProjectService {
...
+                        int facetIndex = facets.findIndexOf { it.name== item.name}
+
+                        if (facetIndex >= 0)
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/services/au/org/ala/biocollect/merit/ProjectService.groovy]: Handled

File: grails-app/assets/javascripts/projects.js [hunk 7]
```diff
@@ -980,6 +981,26 @@ function ProjectViewModel(project, isUserEditor) {
         'Other'
     ];
 
+    self.transients.hasPublishedProjectActivities = false;
+
+    self.checkPublishedProjectActivities = function (activities) {
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/assets/javascripts/projects.js]: Renamed

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 8]
```diff
@@ -611,12 +639,31 @@ class ProjectController {
                 facets = searchService.standardiseHistogramFacets(facets, histogramFacetConfig)
             }
 
-
             // if facet is provided by client do not add special facets
             if(!params.facets){
                 facets = projectService.addSpecialFacets(facets)
             }
 
+            def user = userService.getUser()
```
[the remark this answers is not in the document]
Comment (chamaliperera512) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy]: Done

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 9]
```diff
@@ -684,6 +664,38 @@ class ProjectController {
         render( text: [ projects:  projects, total: searchResult.hits?.total?:0, facets: facets ] as JSON );
     }
 
+    def removeFacetsFromProjectFinderPage(List facets) {
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:667]: make this method private.

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 10]
```diff
@@ -684,6 +664,38 @@ class ProjectController {
...
+
+        def user = userService.getUser()
+        boolean isAlaAdmin = userService.userIsAlaAdmin()
+        boolean isUserPage = params.getBoolean('isUserPage', false)
+
+        // if the user is not logged in or if the user is not AlaAdmin and not in user page, remove the facets specified
+        // in config from the Project Finder Page
+        if (!user || (!isAlaAdmin && !isUserPage)) {
+            for (int i = 0; i < facetsToRemove.size(); i++) {
+                int index = facets.findIndexOf { it == facetsToRemove[i] }
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:678]: you can replace this code with `facets.remove(facetsToRemove[i])`

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 11]
```diff
@@ -684,6 +664,38 @@ class ProjectController {
...
+
+                if (index >= 0)
+                    facets.remove(index)
+            }
+        }
+
+        String facetStr = "";
+
+        if (facets.size() > 0) {
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:687]: you can replace this code with `facets.join(',')`

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy (hunk 9 above)
Comment (temi) [reply grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:667]: Also add some comments to describe what this method does.

**Review by temi — CHANGES_REQUESTED** [review]

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 12]
```diff
@@ -664,33 +664,30 @@ class ProjectController {
...
         List facetsToRemove = grailsApplication.config.lists.facetsToRemoveFromProjectFinderPage
 
         def user = userService.getUser()
         boolean isAlaAdmin = userService.userIsAlaAdmin()
         boolean isUserPage = params.getBoolean('isUserPage', false)
 
-        // if the user is not logged in or if the user is not AlaAdmin and not in user page, remove the facets specified
-        // in config from the Project Finder Page
         if (!user || (!isAlaAdmin && !isUserPage)) {
             for (int i = 0; i < facetsToRemove.size(); i++) {
```
Comment (temi) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:679]: you can remove the whole block and add `facets.removeAll(facetsToRemove)`

**Review by temi — CHANGES_REQUESTED** [review]

**Review by temi — APPROVED** [review]
Looks good

**Pull request merged**